kernel/interrupts/src/aarch64/mod.rs

```diff
@@ -29,6 +29,11 @@ static GIC: MutexIrqSafe<Option<ArmGic>> = MutexIrqSafe::new(None);
 // aarch64 manuals define the default timer IRQ number to be 30.
 pub const CPU_LOCAL_TIMER_IRQ: InterruptNumber = 30;
 
+/// The IRQ/IPI number for TLB Shootdowns
+///
+/// This is arbitrarily defined (range: 0..16)
+pub const TLB_SHOOTDOWN_IPI: InterruptNumber = 2;
```

Comment:
does this need to be fixed/hardcoded? on x86, we have a function that will basically "allocate"/pick a free IRQ number for a new handler (see `register_msi_interrupt` for the basic logic) and then return the IRQ number that it chose. So, you can do this in the function that you currently call `init_ipi`, and then store that IRQ number using `call_once()` after you redefine this as something like:
```rust
pub const TLB_SHOOTDOWN_IPI: Once<InterruptNumber> = Once::new();
```

I think doing something like that here is best, as won't we have devices that may need a specific IRQ number later? (for ex, on x86, ATA disks, keyboard, mouse, etc all have fixed IRQ numbers that cannot be chosen by software).

FWIW, i think this is currently hardcoded on x86 too, but we don't actually use a custom IPI, we use the NMI handler.

Reply:
I'm not sure that peripheral IRQs can land in this range:
- 0..16 is for Software Generated Interrupts = IPIs
- 16..32 is for Local Peripheral Interrupts = the cpu-local timer, for instance
- 32..1024 is for Shared Peripheral Interrupts = PCIe devices for instance

(GIC spec)

So... in this SGI-only range, there shouldn't be any unpredictable user of these numbers 0..16.
But that doesn't prevent us from doing what you said; With this "IRQ number layout" in mind, do you still think it's desirable?